authorization/assignment/role_assignment_model_service.go

```diff
@@ -0,0 +1,136 @@
+type GormRoleAssignmentModelService struct {
+	db           *gorm.DB
+	repositories repositories.Repositories
+}
+
+// ListByResource lists role assignments of a specific resource.
+func (r *GormRoleAssignmentModelService) ListByResource(ctx context.Context, resourceID string) ([]role.IdentityRole, error) {
+	defer goa.MeasureSince([]string{"goa", "db", "identity_role", "list"}, time.Now())
+	var identityRoles []role.IdentityRole
+
+	r.db = r.db.Debug()
+	db := r.db.Raw(`WITH RECURSIVE q AS ( 
+		SELECT 
+		  resource_id, parent_resource_id 
+		FROM 
+		  resource 
+		WHERE 
+		  resource_id = ?
+		UNION ALL
+		SELECT 
+		  p.resource_id, p.parent_resource_id
+		FROM 
+		  resource p
+		JOIN q ON 
+		  q.parent_resource_id = p.resource_id)
+	  SELECT 
+		q.parent_resource_id,q.resource_id, ir.identity_role_id, ir.identity_id, r.role_id, r.name 
+	  FROM 
+		identity_role ir, q, role r
+	  WHERE 
+		ir.resource_id = q.resource_id 
+		and ir.role_id = r.role_id`, resourceID)
+
+	rows, err := db.Rows()
+	if err != nil {
+		log.Error(ctx, map[string]interface{}{
+			"resource_id": resourceID,
+			"err":         err,
+		}, "error running custom sql to get identity roles")
+		return identityRoles, err
+	}
+	defer rows.Close()
+
+	columns, err := rows.Columns()
+	columnValues := make([]interface{}, len(columns))
+
+	var ignore interface{}
+	for index := range columnValues {
+		columnValues[index] = &ignore
+	}
+
+	if err != nil {
+		log.Error(ctx, map[string]interface{}{
+			"resource_id": resourceID,
+			"err":         err,
+		}, "error getting columns")
+		return identityRoles, errors.NewInternalError(ctx, err)
+	}
+
+	for rows.Next() {
+		var parentResourceID *string
+		var returnedResourceID string
+		var identityRoleID string
+		var identityID string
+		var roleID string
+		var roleName string
+
+		columnValues[0] = &parentResourceID
+		columnValues[1] = &returnedResourceID
+		columnValues[2] = &identityRoleID
+		columnValues[3] = &identityID
+		columnValues[4] = &roleID
+		columnValues[5] = &roleName
+
+		if err = rows.Scan(columnValues...); err != nil {
+			log.Error(ctx, map[string]interface{}{
+				"resource_id": resourceID,
+				"err":         err,
+			}, "error getting rows")
+			return identityRoles, errors.NewInternalError(ctx, err)
+		}
+		identityRoleIDAsUUID, _ := uuid.FromString(identityRoleID)
+		identityIDAsUUID, _ := uuid.FromString(identityID)
+		roleIDAsUUID, _ := uuid.FromString(roleID)
+
+		ir := role.IdentityRole{
+			IdentityRoleID: identityRoleIDAsUUID,
+			Identity: account.Identity{
+				ID: identityIDAsUUID,
+			},
+			Resource: resource.Resource{
+				ResourceID:       resourceID,
+				ParentResourceID: parentResourceID,
+			},
+			Role: role.Role{
+				RoleID: roleIDAsUUID,
+				Name:   roleName,
+			},
+		}
+		if parentResourceID != nil {
+			ir.Resource.ParentResourceID = parentResourceID
+		}
+		identityRoles = append(identityRoles, ir)
+	}
+	return identityRoles, nil
```

Comment:
Are you returning all available identity_roles for all identities for this resource or I'm missing something?

Reply:
Yes. Umm. was it just supposed to be for a specific user? I had the github collaborators page in mind where I could see all of us :)

Reply:
From the doc (mentioned in the issue):
"This endpoint is used to list the roles that have been assigned to a **user or team** for a specific resource.  It includes roles that have been assigned explicitly for a resource, in addition to roles that have been inherited from a parent resource."

It should be another endpoint to list the space/area collaborators/members. Not this one if I understand correctly. I don't see much value in the endpoint which list all the identities-roles for the resource. As we agreed the identity-role relation is our implementation detail. We need it to track relations between identities and roles inside Auth.
@sbryzak please clarify if I'm wrong.

Reply:
but the sample response has different assignees .
```
HTTP/1.1 200 OK
...
[{
   "identifier":"9XH4-BN3Y"
   "role_id":"JA7X-PV9Z",
   "role_name":"collaborator",
   "assignee_id":"1111-2222-3333-4444-AAAA-BBBB",
   "assignee_name":"john_smith",
   "assignee_type":"user",
   "inherited":"false"
},
{
   "identifier":"9XH4-BN7J"
   "role_id":"P9A5-MX3H",
   "role_name":"administrator",
   "assignee_id":"9999-8888-7777-6666-ZZZZ-XXXX",
   "assignee_name":"operations_team",
   "assignee_type":"team",
   "inherited":"true",
   "inherited_from":"P3Y8-17WB"
}]
```
A bit confused here - 
`/resource/:resourceID/roles/members` , `/resource/:resourceID/roles/collaborators` seems like very useful endpoints to me similar to https://github.com/fabric8-services/fabric8-auth/settings/collaboration

Are you saying this PR should have an endpoint to retrieve list of roles associated with me , for a specific resource?

Reply:
/resource/:resourceID/roles/members , /resource/:resourceID/roles/collaborators seems to be useful to me too, but this endpoint is different.

> Are you saying this PR should have an endpoint to retrieve list of roles associated with me , for a specific resource?

Yes, this is how I understand it. So, the sample response looks wrong to me.
@sbryzak please clarify what you meant here :)

Reply:
> /resource/:resourceID/roles/members , /resource/:resourceID/roles/collaborators seems to be useful to me too, but this endpoint is different.

`/api/resources/:id/roles/assigned` is an endpoint which returns all users who have *any* role assignment irrespective of what the role is - similar to Github's `GET /repos/:owner/:repo/collaborators` https://developer.github.com/v3/repos/collaborators/



Are you implying that ` /resource/:resourceID/roles/members` , /resource/:resourceID/roles/collaborators` are endpoints which return users who are assigned the named role 'members', 'collaborators' respectively ?

Reply:
OK, I read the doc again and I think I got confused the first time.
I think you are right and this endpoint is just about listing all users and teams which have any role in the resource.

But now I have another concern... what if some user have multiple roles (because of inherited resources)? Will we list all his/her roles? Shouldn’t we return the most powerful role only, so, there is only one entry per user/team/group? We could resolve it by analyzing the scopes of overlapping resources.

Reply:
Cool. 

> But now I have another concern... what if some user have multiple roles (because of inherited resources)? Will we list all his/her roles? 

Ah yeah, this API consciously does that I think, because we have the `inherited:true` and `parent_resource_id` in the response

How do we determine if the scope is the most powerful ? Let me read the doc again..

Reply:
We don't know what scope is most powerful because we don't know what the scope mean. It's up to resource service to interpret them. But we could try to resolve the most powerful **role** by analyzing the super/sub set of scopes from these roles... It can be tricky though. @sbryzak WDYT?